File: tests/utils/generate_synthetic_labels.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

#include <iostream>
#include <random>
#include <boost/program_options.hpp>

#include "utils.h"

namespace po = boost::program_options;

int main(int argc, char** argv) {
  std::string output_file;
  _u64        num_labels, num_points;

  try {
    po::options_description desc{"Arguments"};

    desc.add_options()("help,h", "Print information on arguments");
    desc.add_options()("output_file,O",
                       po::value<std::string>(&output_file)->required(),
                       "Filename for saving the label file");
    desc.add_options()("num_points,N",
                       po::value<uint64_t>(&num_points)->required(),
                       "Number of points in dataset");
    desc.add_options()("num_labels,L",
                       po::value<uint64_t>(&num_labels)->required(),
                       "Number of unique labels, up to 100");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    if (vm.count("help")) {
      std::cout << desc;
      return 0;
    }
    po::notify(vm);
  } catch (const std::exception& ex) {
    std::cerr << ex.what() << '\n';
    return -1;
  }

  if (num_labels > 100) {
    std::cerr << "Error: num_labels must be 100 or less" << '\n';
    return -1;
  }

  if (num_points <= 0) {
    std::cerr << "Error: num_points must be greater than 0" << '\n';
    return -1;
  }

  std::cout << "Generating synthetic labels for " << num_points
            << " points with " << num_labels << " unique labels" << '\n';

  try {
    std::ofstream outfile(output_file);
    if (!outfile.is_open()) {
      std::cerr << "Error: could not open output file " << output_file << '\n';
      return -1;
    }

    for (int i = 0; i < num_points; i++) {
      bool label_written = false;
      for (int j = 1; j <= num_labels; j++) {
        // 50% chance to assign each label
        if (rand() > (RAND_MAX / 2)) {
          if (label_written) {
            outfile << ',';
          }
          outfile << j;
          label_written = true;
        }
      }

      if (i < num_points - 1) {
        outfile << '\n';
      }
    }
    outfile.close();

    std::cout << "Labels written to " << output_file << '\n';

  } catch (const std::exception& ex) {
    std::cerr << "Label generation failed: " << ex.what() << '\n';
    return -1;
  }

  return 0;
}

File: diskann-tools/src/bin/generate_synthetic_labels.rs
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT license.
 */
use clap::Parser;
use diskann_providers::utils::generate_synthetic_labels_utils::generate_labels;
use diskann_tools::utils::{CMDResult, CMDToolError};
use tracing::{error, info};

#[derive(Debug, Parser)]
struct GenerateSyntheticLabelsArgs {
    /// Filename for saving the label file
    #[arg(long = "output_file", required = true)]
    pub output_file: String,

    /// Number of vectors
    #[arg(long = "npts", short = 'N', required = true)]
    pub number_of_vectors: u64,

    /// Number of labels
    #[arg(long = "nlbls", short = 'L', required = true)]
    pub number_of_labels: u32,

    /// Distribution function for labels random/zipf/one_per_point, defaults to random
    #[arg(long = "ndt", short = 'D', required = true, default_value = "random")]
    pub distribution_type: String,
}

fn main() -> CMDResult<()> {
    let args: GenerateSyntheticLabelsArgs = GenerateSyntheticLabelsArgs::parse();

    if args.number_of_labels > 5000 {
        return Err(CMDToolError {
            details: "Error: num_labels must be 5000 or less".to_string(),
        });
    }

    if args.number_of_vectors == 0 {
        return Err(CMDToolError {
            details: "Error: num_points must be greater than 0".to_string(),
        });
    }

    info!(
        "Generating synthetic labels for {} points with {} unique labels.",
        args.number_of_vectors, args.number_of_labels
    );

    match generate_labels(
        &args.output_file,
        &args.distribution_type,
        args.number_of_vectors as usize,
        args.number_of_labels,
        &mut diskann_providers::utils::create_rnd_from_seed(42),
    ) {
        Ok(_) => {
            info!("Successfully generated labels");
            Ok(())
        }
        Err(err) => {
            error!("Label generation failed: {:?}", err);
            Err(err.into())
        }
    }
}
